fix(linter/prefer-array-flat-map): error for `.flat(1.0)`

overlookmotel opened the pull request:
Previously this rule did not produce an error for `.map(f).flat(1.0)`. This PR fixes that.

Also round down the argument and produce errors for e.g. `.map(f).flat(1.9)`. This follows ECMA spec (see comments in code), and I've also verified that all the weird test cases match behavior of `.flat(...)` in NodeJS.

overlookmotel commented:
* **#12360** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12360?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12360?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12360).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Duplicate of #12358.

@camc314 We both worked on the same thing! There's a very slight difference between our 2 fixes. I *think* mine is correct (using `== 1.0` comparison).

camc314 commented:
### Merge activity

* **Jul 17, 12:51 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(linter/prefer-array-flat-map): error for `.flat(1.0)` (#12360)

Previously this rule did not produce an error for `.map(f).flat(1.0)`. This PR fixes that.

Also round down the argument and produce errors for e.g. `.map(f).flat(1.9)`. This follows ECMA spec (see comments in code), and I've also verified that all the weird test cases match behavior of `.flat(...)` in NodeJS.
